"panic: (*vector.Const)" in vector runtime query against ~1-billion values

## tl;dr

```
$ super -version
Version: v1.18.0-426-g196d094e7

$ SUPER_VAM=1 super -c "
SELECT
    commit.collection AS event,
    EXTRACT(HOUR FROM time(time_us*1000)) as hour_of_day,
    count() AS count
FROM 'bluesky-196d094.csup'
WHERE kind = 'commit'
  AND commit.operation = 'create'
  AND commit.collection in ['app.bsky.feed.post', 'app.bsky.feed.repost', 'app.bsky.feed.like']
GROUP BY event, hour_of_day
ORDER BY hour_of_day, event;"
panic: (*vector.Const) 0xc0024da320

goroutine 15 [running]:
github.com/brimdata/super/runtime/vam/expr/function.date_time_hour({0x23a4a50?, 0xc0024da320?})
        /home/ubuntu/super/runtime/vam/expr/function/datepartfuncs.go:107 +0x2d3
github.com/brimdata/super/runtime/vam/expr/function.(*DatePart).Call(0xc00042c100, {0xc00095c460, 0x2, 0x479285?})
        /home/ubuntu/super/runtime/vam/expr/function/datepart.go:32 +0x1f4
github.com/brimdata/super/vector.Apply(0x8?, 0x2?, {0xc00095c460?, 0xc0036322c0?, 0x1e9d000?})
        /home/ubuntu/super/vector/apply.go:17 +0xa3
github.com/brimdata/super/vector.Apply(0x1, 0xc001cd1d60, {0xc000113aa0?, 0xc001cd1d68?, 0xc000359b80?})
        /home/ubuntu/super/vector/apply.go:23 +0x10e
github.com/brimdata/super/runtime/vam/expr.(*Call).Eval(0xc0005f94f0, {0x23a45a0, 0xc003632040})
        /home/ubuntu/super/runtime/vam/expr/eval.go:39 +0xf4
github.com/brimdata/super/runtime/vam/op/aggregate.(*Aggregate).Pull(0xc0004891e0, 0x0?)
        /home/ubuntu/super/runtime/vam/op/aggregate/aggregate.go:72 +0x41a
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0xc0007aef40)
        /home/ubuntu/super/runtime/vam/op/combine.go:110 +0x3c
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 12
        /home/ubuntu/super/runtime/vam/op/combine.go:42 +0x35
```

## Details

Repro is with super commit 196d094 and a CSUP file made from a cleaned up copy of the ~1-billion Bluesky events in the JSONBench data set.

This problem doesn't trigger with any of the sets of 100-million events at a time. Thankfully @nwt has eyeballed the stack trace and thinks he sees a fix, so I'll plan to test out a branch on my AWS instance where I have the full CSUP data file ready to go.

--- a/runtime/vam/expr/function/datepartfuncs.go
+++ b/runtime/vam/expr/function/datepartfuncs.go
@@ -19,6 +19,10 @@ func date_time_dayofweek(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Weekday())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Weekday()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_dayofweek(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -44,6 +48,10 @@ func date_time_day(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Day())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Day()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_day(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -69,6 +77,10 @@ func date_time_dow(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Weekday())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Weekday()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_dow(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -94,6 +106,10 @@ func date_time_hour(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Hour())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Hour()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_hour(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -119,6 +135,10 @@ func date_time_microseconds(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Second()*1e6 + nano.Ts(v).Time().Nanosecond()/1e3)
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Second()*1e6 + nano.Ts(v).Time().Nanosecond()/1e3))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_microseconds(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -144,6 +164,10 @@ func date_time_milliseconds(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Second()*1e3 + nano.Ts(v).Time().Nanosecond()/1e6)
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Second()*1e3 + nano.Ts(v).Time().Nanosecond()/1e6))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_milliseconds(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -169,6 +193,10 @@ func date_time_minute(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Minute())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Minute()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_minute(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -194,6 +222,10 @@ func date_time_month(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Month())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Month()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_month(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -219,6 +251,10 @@ func date_time_second(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Second())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Second()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_second(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
@@ -244,6 +280,10 @@ func date_time_year(vec vector.Any) vector.Any {
 			out[i] = int64(nano.Ts(v).Time().Year())
 		}
 		return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+	case *vector.Const:
+		v := vec.Value().Int()
+		val := super.NewInt64(int64(nano.Ts(v).Time().Year()))
+		return vector.NewConst(val, vec.Len(), vec.Nulls)
 	case *vector.Dict:
 		out := date_time_year(vec.Any).(*vector.Int)
 		return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)

--- a/runtime/vam/expr/function/gendatepartfuncs.go
+++ b/runtime/vam/expr/function/gendatepartfuncs.go
@@ -32,16 +32,20 @@ case *vector.View:
 	out := make([]int64, len(index))
 	for i, idx := range index {
 		v := inner.Values[idx]
-		out[i] = %s
+		out[i] = %[1]s
 	}
 	return vector.NewInt(super.TypeInt64, out, inner.Nulls.Pick(index))
+case *vector.Const:
+	v := vec.Value().Int()
+	val := super.NewInt64(%[1]s)
+	return vector.NewConst(val, vec.Len(), vec.Nulls)
 case *vector.Dict:
-    out := %s(vec.Any).(*vector.Int)
+	out := %[2]s(vec.Any).(*vector.Int)
 	return vector.NewDict(out, vec.Index, vec.Counts, vec.Nulls)
 case *vector.Int:
 	out := make([]int64, vec.Len())
 	for i, v := range vec.Values {
-		out[i] = %s
+		out[i] = %[1]s
 	}
 	return vector.NewInt(super.TypeInt64, out, vec.Nulls)
 default:
@@ -61,7 +65,7 @@ func main() {
 	fmt.Fprintln(&buf, ")")
 	for _, fn := range funcs {
 		fmt.Fprintf(&buf, "func %s(vec vector.Any) vector.Any {\n", funcName(fn.name))
-		fmt.Fprintf(&buf, datePartSwitch, fn.expr, funcName(fn.name), fn.expr)
+		fmt.Fprintf(&buf, datePartSwitch, fn.expr, funcName(fn.name))
 		fmt.Fprintf(&buf, "}\n\n")
 	}
 	buf.WriteString("var datePartFuncs = map[string]func(vector.Any)vector.Any {\n")

